handlers/article.go

```diff
@@ -0,0 +1,437 @@
+	}
+
+	err = query.Find(&articles).Error
+
+	if tags, ok := queryParams["tag"]; ok {
+		articles, err = h.DB.GetAllArticlesWithTag(tags[0])
+	}
+
+	if authorBy, ok := queryParams["author"]; ok {
+		articles, err = h.DB.GetAllArticlesAuthoredBy(authorBy[0])
+	}
+
+	if favoritedBy, ok := queryParams["favorited"]; ok {
+		articles, err = h.DB.GetAllArticlesFavoritedBy(favoritedBy[0])
+	}
+
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+
+	w.Header().Set("Content-Type", "application/json")
+
+	if len(articles) == 0 {
+		json.NewEncoder(w).Encode(ArticlesJSON{})
+		return
+	}
+
+	var u = &models.User{}
+	if claim, _ := h.JWT.CheckRequest(r); claim != nil {
+		u, err = h.DB.FindUserByUsername(claim.Username)
+	}
+
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusNotFound)
+		return
+	}
+
+	var articlesJSON ArticlesJSON
+	for i := range articles {
+		a := &articles[i]
+		articlesJSON.Articles = append(articlesJSON.Articles, h.buildArticleJSON(a, u))
+	}
+
+	articlesJSON.ArticlesCount = len(articles)
+
+	json.NewEncoder(w).Encode(articlesJSON)
+}
+
+// createArticle handle POST /api/articles
+func (h *Handler) createArticle(w http.ResponseWriter, r *http.Request) {
+	var body struct {
+		Article struct {
+			Title       string   `json:"title"`
+			Description string   `json:"description"`
+			Body        string   `json:"body"`
+			TagsList    []string `json:"tagsList"`
+		} `json:"article"`
+	}
+
+	if err := json.NewDecoder(r.Body).Decode(&body); err != nil {
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+
+	defer r.Body.Close()
+
+	u := r.Context().Value(contextKeyCurrentUser).(*models.User)
+
+	a := models.NewArticle(body.Article.Title, body.Article.Description, body.Article.Body, u)
+
+	if valid, errs := a.IsValid(); !valid {
+		w.Header().Set("Content-Type", "application/json")
+		w.WriteHeader(http.StatusUnprocessableEntity)
+		errorResponse := errorResponse{Errors: errs}
+		json.NewEncoder(w).Encode(errorResponse)
+		return
+	}
+
+	for _, tagName := range body.Article.TagsList {
+		tag, _ := h.DB.FindTagOrInit(tagName)
+		a.Tags = append(a.Tags, tag)
+	}
+
+	if err := h.DB.CreateArticle(a); err != nil {
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+
+	articleJSON := ArticleJSON{
+		Article: h.buildArticleJSON(a, u),
+	}
+
+	w.Header().Set("Content-Type", "application/json")
+	w.WriteHeader(http.StatusCreated)
+	json.NewEncoder(w).Encode(articleJSON)
+}
+
+// updateArticle handle PUT /api/articles/:slug
+func (h *Handler) updateArticle(w http.ResponseWriter, r *http.Request) {
+	var err error
+	a := r.Context().Value(contextKeyArticle).(*models.Article)
+	u := r.Context().Value(contextKeyCurrentUser).(*models.User)
+
+	if !a.IsOwnedBy(u.Username) {
+		err = fmt.Errorf("You don't have the permission to edit this article")
+		http.Error(w, err.Error(), http.StatusForbidden)
+		return
+	}
+
+	var body map[string]map[string]interface{}
```

Comment:
Hmm Good question I thinks I like the way Go check the presences of key in a map
```
if value, ok: aMap["key"]; ok {
...
}
```
and also because the fields are optional for update